### ISSUE TITLE
What is maskmode(in models.lua defineG_shiftNet) used for?

### ISSUE THREAD

**Issue by zzzzyyang** [Zhaoyi-Yan/Shift-Net#10]
Dear author, thank you for your excellent work.
I can't understand the code in models.lua
https://github.com/Zhaoyi-Yan/Shift-Net/blob/b4bb2d19e33320efd26b2bd94187b213bce8a996/models.lua#L49  
https://github.com/Zhaoyi-Yan/Shift-Net/blob/b4bb2d19e33320efd26b2bd94187b213bce8a996/models.lua#L13 
Could you please  tell me why it is needed?

**Comment by Zhaoyi-Yan** [comment]
Hi,
it represents a three-layer convs(1 channel each, with all weights 1/16). It is just used to represent the masked region in feature space. `threshold` can filter out masked region and unmasked region, by  judging whether the correspond value  is larger than `threshold` or not.

**Comment by zzzzyyang** [comment]
> Hi,
> it represents a three-layer convs(1 channel each, with all weights 1/16). It is just used to represent the masked region in feature space. `threshold` can filter out masked region and unmasked region, by judging whether the correspond value is larger than `threshold` or not.

**Comment by zzzzyyang** [comment]
Thank you for the quick reply! It helps a lot!
But i still have some problem with this code,in train.lua https://github.com/Zhaoyi-Yan/Shift-Net/blob/7669232c22d4b16a989ecd8169ec9983215afefe/train.lua#L426 
Is df_dummy the gradient of the guidance loss?why is there zero() function? Could you please explain it ?
thanks!

**Comment by Zhaoyi-Yan** [comment]
https://github.com/Zhaoyi-Yan/Shift-Net/blob/master/lib/innerCos.lua#L51
```lua
function AddCos:updateGradInput(input, gradOutput)
    local gradInput_former = self.criterion:backward(self.input_latent_mask, self.GT_latent_mask)
    gradInput_former:mul(self.strength)
    gradInput_former:add(gradOutput)

    local dummy_mask_grad = torch.Tensor(self.bz, 1, self.h, self.w):typeAs(input):zero()
    self.gradInput = torch.cat({gradInput_former, dummy_mask_grad}, 2)
    return self.gradInput
end
```
Guidance loss is handled in InnerCos.lua, the truth gradients is calculate in this function.
So you need to pass `zero graidents` in.

**Comment by Zhaoyi-Yan** [comment]
Isn't is solved? @zzzzyyang

**Comment by zzzzyyang** [comment]
yes!Thanks
